### ISSUE TITLE
requirements for first release version

### ISSUE THREAD

**Issue by basisbit** [UltraStar-Deluxe/Play#104]
Issue type: Development discussion
### Problem or question to be solved
So far, there exists no build of UltraStar Play for people to download and use. The question now is 
- when to build the next public release of the game and 
- what issues have to be done before we release and
- what other tasks should be done ahead of the release?

### Suggested solution/s
Initially, I planned on often releasing new non-stable versions as long as the game works mostly.

### Your opinion on the topic
So far, there exists no website for this project - I'd consider having one before release to be mandatory. (there exists the usplay.net domain already, but no content)
I guess we will also need some very basic user-targeting documentation.

What are your opinions on these topics?

**Comment by achimmihca** [comment]
I also consider a website mandatory. At least I would add there: A short description about the game and development plans (it is in development and we are looking for people to contribute), download link, and link to contibution guide.

I want to include the open PR for the song-editor. I have the hopes that a strong editor will be a selling point for UltraStar Play. Thus, I plan to write some documentation how to use the editor (e.g. keyboard shortcuts) in the next few days.

> what issues have to be done before we release?

- I think bug #84 should be fixed before a release. Most people will use a different resolution than 800x600.
- For proper game play the note detection needs to be improved. I think the missing delay between mic input and recorded note is a real problem (#86). In #12 you said you earned ~6500 points in USP whereas it was ~9500 in USDX. IMO this is too much a difference. Latency should be a parameter in the MicProfile that can be changed in the UI. You said USDX assums a default latency of 140ms.

**Comment by basisbit** [comment]
Yesterday at a birthday party, I had some people (who are rather good at singing the played songs) play karaoke using UltraStar Play at difficulty easy, and they did not do well, and rather quickly stopped singing to go and do other stuff. So yes, definitely not good enough for a release yet.
From what it looked like, the singer pitch detection happens for beats which are a bunch of beats after the beat which the singers actually did sing to. While 140ms might work for USDX, the perceived latency is much much higher for UltraStar Play.

**Comment by achimmihca** [comment]
Sooo, improving the pitch detection using YIN did not work as I expected.

Instead I suggest to implement something similar to "coyote time", which gives players an advantage (see https://www.youtube.com/watch?v=sLXLlJ7FhJU at 5:20). The goal is to make scoring feel fair and the player feel like a boss.

One could, for example, delay the scoring by N beats. If the current N beats to-be-scored were detected as "wrong", but the previous and the following N beats were measured as "correct", then also handle the current N beats as if they were measured as "correct".
Such an approach is not costly, and could make our pitch detection feel more stable (although in reality it is not). It makes use of the additional knowledge that we have about the expected pitch.
Good players are not irritated by single false detected pitches that destroy their perfect sentence, but it also does no hand-holding (so if you do not hit a single note, then this does not help you).

**Comment by basisbit** [comment]
I'd like to push for a "preview" release in one or two weeks if that is ok for you. There is no need for all the bugs that we know of to be fixed, because we only want to show people something like "hey, this project exists and we are working on cool stuff, using modern tools!" - no need to target non-developers yet, I guess.

I haven't heard anything from @Baklap4 regarding any progress creating a website, so I'll start reading on how to set up a github.io hosted Jekyll website and fill it with some basic content, screenshots and links to the wiki pages.

**Comment by achimmihca** [comment]
Great! I will take a look at #84 sometime this weekend and then it should be fine I guess.

**Comment by achimmihca** [comment]
In my opinion we are ready for a first alpha release (yay!). You might want to test again what we have so far.

BTW: You said something about an upcoming USDX release? Is this still in the works? I could use a new version ;)

**Comment by basisbit** [comment]
It is ready since 6 days ago, so after sone more testing from some friends, we could probably do a release this weekend. There are still some tasks to do: 
- [ ] improve scaling of UI for high resolutions (4k or 2k screens with 16:9 or typical old 1024*786px 4:3 screens should all work)
- [ ] some basic documentation page on how to set up and start the game on the website
- [ ] test linux build and macOS build

I started the process to publish the game on Steam and Google Play store which would also include easy distribution of updates.

**Comment by natrius** [comment]
As it is open source, do you plan to release it on F-Droid store as well? Its an open source store for android. https://f-droid.org/ You can use a special template so its easy to see directly in the app the last changes or the cahngeset of the last update. Should work automatically as well, afaik.

**Comment by basisbit** [comment]
@natrius please create a new issue for that. If you know how to do this, please add a release-documentation/instructions document to the git repository and create a Pull Request. Thanks!

**Comment by achimmihca** [comment]
I'd like to finally create a release with the current status.
I think now is a good time before doing major updates such as the new input system or new UI toolkit.
**Also, I would like more people to get interested in the project.**

**What is missing for a release?**

- The "Alpha" project names some criteria: https://github.com/UltraStar-Deluxe/Play/projects
    - "Supports very basic features: loading progress bar, song selection, see current lyrics, song audio playback, song video playback, pitch detection, grade singer performance, settings"
        - check
    - "Auto-update or update notifier"
        - exists and should be fixed with #203
    - "Logging"
        - There is logging. IMO it is ok for a first release.
        - what exactly is desired here? Something like apache-commons logging? I think the Debug.Log statement of Unity is pretty verbose. I guess this could be improved in a new issue. But it is not required for a first release, right?
    - "Includes at least one test-song"
        - I think the download-songs feature is good enough.
    - "Has some unit tests and/or integration tests"
        - check

IMO we met all requirements of the "Alpha" project.

- We also have a website: https://usplay.net/
    - Who has access and can update the text once the release is ready?
- Which version name do we want to use?
    - The [VERSION.txt](https://github.com/UltraStar-Deluxe/Play/blob/master/UltraStar%20Play/Assets/VERSION.txt) file needs some value for "release" that makes comparing with future versions possible.
        - This value is used by the NewVersionChecker and the format should not change. So I suggest something natural based on time.
        - I would do `year-month-day` or `year-month` (like Ubuntu does) or `year-number` (like Unity does)
        - Note that the user might still see another, more "human" version name

> some basic documentation page on how to set up and start the game on the website

@basisbit what exactly do you have in mind?

- For me it is: go to settings, add player profiles, add mic profiles, add songs
    - I suggest to also document this here: https://github.com/UltraStar-Deluxe/Play/wiki/First-Steps
    - Finding / adding songs is explained here: https://github.com/UltraStar-Deluxe/Play/wiki/Songs

> test linux build and macOS build

- I suggest to do a pre-release version for testing (pretty much now, after PR have been merged).
    - If everything is ok with pre-release, we can finalize it and update website with download link.
- Question: The travis build will at some point remove the uploaded artifacts. I guess, for a final release we have to copy the files to a different location with a different download url?

> I started the process to publish the game on Steam and Google Play store which would also include easy distribution of updates.

- This is nice. However, I don't think we need Steam (for a first release). I would start small and put a download link somewhere on the website.
- At the moment, I consider Android nice-to-have. What is more important IMO is a release for desktop.
    - I would not release Android before moving to a new Input system. And I would like a release before changing the input system.

**Comment by achimmihca** [comment]
I added logging with #216 and I added [first steps](https://github.com/UltraStar-Deluxe/Play/wiki/First-Steps) documentation.

So, let's do a pre-release for testing now?

**Comment by achimmihca** [comment]
I just created a release with the current status (v0.2.1).
For now I will close this issue.

Will create issues for future development soon.
